File: 1612-avoid-flood-in-the-city/avoid-flood-in-the-city.java
class Solution {
    public int[] avoidFlood(int[] rains) {
        int n = rains.length;
        int[] result = new int[n];
        Arrays.fill(result, 1);
        
        HashMap<Integer, Integer> full = new HashMap<>(); 
        TreeSet<Integer> dryDays = new TreeSet<>();      
        
        for (int i = 0; i < n; i++) {
            int lake = rains[i];
            
            if (lake == 0) {
                dryDays.add(i); 
            } else {
                result[i] = -1; 
                
                if (full.containsKey(lake)) {
                    Integer dryDay = dryDays.higher(full.get(lake));
                    if (dryDay == null) {
                        return new int[0]; 
                    }
                    
                    result[dryDay] = lake; 
                    dryDays.remove(dryDay); 
                }
                
                full.put(lake, i); 
            }
        }
        
        return result;
    }
}

File: 1612-avoid-flood-in-the-city/avoid-flood-in-the-city.cpp
class Solution {
public:
    vector<int> avoidFlood(vector<int>& rains) {
        int n = rains.size();
        vector<int> ans(n, 1);
        unordered_map<int,int> last; 
        set<int> dryDays; 
        
        for (int i = 0; i < n; i++) {
            int lake = rains[i];
            if (lake == 0) {
                dryDays.insert(i); 
            } else {
                ans[i] = -1;
                if (last.count(lake)) {
                    auto it = dryDays.upper_bound(last[lake]);
                    if (it == dryDays.end()) return {}; 
                    ans[*it] = lake; 
                    dryDays.erase(it);
                }
                last[lake] = i;
            }
        }
        return ans;
    }
};
